### PR TITLE
[MIG] sale_order_product_recommendation: Migration to 12.0

### PR DESCRIPTION
**Pull request by chienandalu**
Standard migrations

cc @Tecnativa TT20130

### CODE REVIEW

File: sale_order_product_recommendation/__manifest__.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 {
     "name": "Sale Order Product Recommendation",
     "summary": "Recommend products to sell to customer based on history",
-    "version": "11.0.1.0.1",
+    "version": "12.0.1.0.1",
```
Comment (pedrobaeza) [inline sale_order_product_recommendation/__manifest__.py]: 12.0.1.0.0

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Tests are failing

**Review by yajo — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Travis still red

**Comment by chienandalu** [comment]
Looks like there's some interaction with another module of the report, my bet goes for sale_automatic_workflow changing delivered_qty behavior.

**Comment by pedrobaeza** [comment]
> Looks like there's some interaction with another module of the report, my bet goes for sale_automatic_workflow changing delivered_qty behavior.

I don't think so. I would say the test is not well programmed, trying to use `[2]` index while there aren't that product lines in the test data. Why don't you filter that lines by product instead for doing the test more resilient?

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        required=True,
+        help="The less, the faster they will be found.",
+    )
+    last_compute = fields.Char()
+
+    @api.model
+    def _default_order_id(self):
+        return self.env.context.get("active_id", False)
+
+    @api.multi
```
Comment (sergio-teruel) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]: Remove multi

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        """Generate lines according to context sale order."""
+        last_compute = '{}-{}-{}'.format(
+            self.id, self.months, self.line_amount)
+        # Avoid execute onchange as times as fields in api.onchange
+        # ORM must control this?
+        if self.last_compute == last_compute:
+            return
+        self.last_compute = last_compute
+        start = datetime.now() - timedelta(days=self.months * 30)
+        start = fields.Datetime.to_string(start)
```
Comment (sergio-teruel) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]: In V12.0 do you need convert to string??

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            # Use a new in-memory line otherwise
+            so_line = so_lines.new({
+                "order_id": self.order_id.id,
+                "product_id": wiz_line.product_id.id,
+                "sequence": sequence,
+            })
+            so_line.product_id_change()
+            so_line.product_uom_qty = wiz_line.units_included
+            so_line.product_uom_change()
+            so_lines |= so_line
```
Comment (sergio-teruel) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]: For performance is better store createed line id and finally do a browse with all ids...

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
Fix travis...

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py (hunk 4 above)
Comment (chienandalu) [reply sale_order_product_recommendation/wizards/sale_order_recommendation.py]: Even with on memory records? Changed anyway

**Comment by chienandalu** [comment]
> I would say the test is not well programmed, trying to use `[2]` index while there aren't that product lines in the test data. Why don't you filter that lines by product instead for doing the test more resilient?

I'd say the problem comes from computing the delivered qtys. Anyway, I changed the tests as to avoid sorting problems like that

**Comment by chienandalu** [comment]
Green at last. It was indeed a problem with the computation of delivered qtys in the test. Setting it wasn't enough and finally I had to force the product to be services so the stock methods didn't interfere...

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**